File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <bits/stdc++.h>
using namespace std;

int N,ans;
bool col[31],diag1[31],diag2[31];

void solve(int cnt) {
    if(cnt==N) {
        ans++;
        return;
    }
    for(int i=0; i<N; i++) {
        if(col[i]||diag1[i+cnt]||diag2[cnt-i+N-1]) continue;
        col[i]=1;
        diag1[i+cnt]=1;
        diag2[cnt-i+N-1]=1;
        solve(cnt+1);
        col[i]=0;
        diag1[i+cnt]=0;
        diag2[cnt-i+N-1]=0;
    }
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin >> N;
    solve(0);
    cout << ans;
}

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class Main {
	static int N,ans;
	static boolean[] row = new boolean[31];
	static boolean[] diag1 = new boolean[31];
	static boolean[] diag2 = new boolean[31];
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	public static void main(String[] args) throws IOException {
		N = Integer.parseInt(br.readLine());
		solve(0);
		System.out.println(ans);
	}
	
	static void solve(int cnt) {
		if(cnt==N) {
			ans++;
			return;
		}
		for(int i=0; i<N; i++) {
			if(row[i]||diag1[i+cnt]||diag2[cnt-i+N-1]) continue;
			row[i]=true;
			diag1[i+cnt]=true;
			diag2[cnt-i+N-1]=true;
			solve(cnt+1);
			row[i]=false;
			diag1[i+cnt]=false;
			diag2[cnt-i+N-1]=false;
		}
	}
}